### PR TITLE
feat(vibecuda_softmax): add SM103 softmax backend

### PR DESCRIPTION
**Pull request by Edenzzzz**
## Description

This PR adds an SM103 softmax kernel generated by VibeCUDA. On NVIDIA B300
SXM6 AC, it achieves **1.3494x arithmetic-mean and 1.3371x geometric-mean
speedup** over the optimized CAKE softmax from
[PR 4282](https://github.com/flashinfer-ai/flashinfer/pull/4282) at
`f0edac69e1c4d299c14fa95b7a0073a172203cf9` across 41/41 precision-passing
workloads. Because that CAKE baseline is not yet in `main`, the canonical
repository benchmark also measures **1.9486x arithmetic-mean and 1.8989x
geometric-mean speedup** over the current default FlashInfer softmax across
40/40 precision-passing workloads. It adds a CUDA backend to
`flashinfer.sampling.softmax(..., backend="vibecuda")` while preserving the
existing default backend.

The CAKE comparison covers the PR 4282 40-row FP32 batch/vocabulary matrix plus
its scalar-temperature PDL row (`batch=64`, `vocab=32000`, `temperature=1.0`,
`enable_pdl=True`). Performance claims apply to SM103 only; SM100 was not
profiled.

### Public contract

- `flashinfer.sampling.softmax(logits, temperature, enable_pdl,
  backend="vibecuda")` returns a fresh FP32 output.
- Scalar and per-row temperatures, `-inf` inputs, and PDL are supported.
- The VibeCUDA JIT module has build targets for SM100 and SM103; this PR's
  optimization and performance validation target is SM103 only.
- Unknown backend names fail explicitly rather than silently falling back.

### Architecture and source provenance

- Profiled target: SM103 on NVIDIA B300 SXM6 AC.
- Target base commit: `ff22228d2fa144e9ac6a0d841f2e9ba767ba0f0a`.
- Fast baseline: optimized CAKE softmax from
  [PR 4282](https://github.com/flashinfer-ai/flashinfer/pull/4282) at
  `f0edac69e1c4d299c14fa95b7a0073a172203cf9`.
- Current-main baseline: the default FlashInfer sampling softmax at the target
  base commit.
- Device header: 58,630 bytes, SHA-256
  `4788393305e4ef946489055e5cf1407e658b10ae6bae4f4bef6c1dcaf046c8f1`.

## Validation

GPU checks on SM103:

- `PYTHONPATH=$PWD python -m pytest tests/utils/test_sampling.py -k
  'softmax_vibecuda' -q`: **541 passed**, 1,911 deselected.
- Both the CAKE softmax and VibeCUDA pass **41/41** rows against
  `torch.softmax` at `atol=1e-5`.
- The CAKE route probe confirms its custom SM103 routes execute on every
  denominator row; the PDL row executes route 4.

## Performance

Protocol: repository `bench_gpu_time` CUDA-event timing, per-iteration cold-L2
flush, no CUDA Graph, 100 ms warmup, 1,000 ms measurement arm, and per-row
median. Candidate and each named baseline use identical FP32 inputs. All
advertised rows pass precision.

### Versus the optimized CAKE softmax (PR 4282, not yet in main)

| Workload | CAKE latency (ms) | VibeCUDA latency (ms) | Speedup |
|---|---:|---:|---:|
| `b1, v64000, temperature=None, PDL off` | 0.0113 | 0.0092 | 1.221x |
| `b16, v256000, temperature=None, PDL off` | 0.0195 | 0.0154 | 1.269x |
| `b1024, v128000, temperature=None, PDL off` | 0.2898 | 0.1712 | 1.693x |
| `b64, v32000, temperature=1.0, PDL on` | 0.0114 | 0.0113 | 1.011x |

- **Arithmetic mean:** 1.3494x over 41/41 rows
- **Geometric mean:** 1.3371x over 41/41 rows
- **Minimum / maximum:** 1.0000x at `b128, v32000` / 1.7286x at
  `b128, v256000`

### Versus the current default FlashInfer softmax in main

| Workload | Default latency (ms) | VibeCUDA latency (ms) | Speedup |
|---|---:|---:|---:|
| `b1, v64000, temperature=None, PDL off` | 0.0175 | 0.0092 | 1.89x |
| `b16, v256000, temperature=None, PDL off` | 0.0399 | 0.0154 | 2.59x |
| `b512, v128000, temperature=None, PDL off` | 0.1628 | 0.0912 | 1.79x |
| `b1024, v128000, temperature=None, PDL off` | 0.3122 | 0.1731 | 1.80x |

- **Arithmetic mean:** 1.9486x over 40/40 rows
- **Geometric mean:** 1.8989x over 40/40 rows
- **Minimum / maximum:** 1.2145x at `b256, v32000` / 3.6540x at
  `b1, v256000`

### Direct verification commands

```bash
PYTHONPATH=$PWD python3 -m pytest tests/utils/test_sampling.py -k "test_softmax_vibecuda" -q
PYTHONPATH=$PWD python3 benchmarks/bench_softmax.py
```

The canonical benchmark command above directly prints the full current-main
baseline/VibeCUDA table and aggregate. The separately reported CAKE result was
measured against the pinned unmerged PR 4282 revision with the same repository
timing and precision protocol; it is not mixed into the current-main aggregate.

## References

- [PR 4282: optimized CAKE softmax](https://github.com/flashinfer-ai/flashinfer/pull/4282)
  at `f0edac69e1c4d299c14fa95b7a0073a172203cf9`
- [Canonical benchmark source](https://github.com/Edenzzzz/flashinfer/blob/08042164/benchmarks/bench_softmax.py)
- [Primary tests](https://github.com/Edenzzzz/flashinfer/blob/08042164/tests/utils/test_sampling.py)
- [CAKE kernel tracker, issue 4254](https://github.com/flashinfer-ai/flashinfer/issues/4254)

Related to #4254.

## Checklist

- [x] Repository tests cover the public API, temperatures, PDL, `-inf` inputs,
  and backend validation.
- [x] Candidate and each named baseline use matched precision and benchmark
  protocols.
- [x] All advertised workloads pass the precision contract.
- [x] `git diff --check` passes.

## Reviewer notes

The primary review surfaces are the public dispatch in `flashinfer/sampling.py`,
the JIT/AOT registration, the binding and launcher, the device header, tests,
and the canonical `benchmarks/bench_softmax.py` comparison. No SM100 performance
claim is made.

## Summary by CodeRabbit

- **New Features**
  - Added an optional VibeCUDA softmax backend for supported NVIDIA SM100-class GPUs.
  - Softmax now supports backend selection, scalar or per-row temperature scaling, and optional programmatic dependent launch (PDL) execution.
  - Added automatic build integration for supported GPU architectures.
- **Benchmarking**
  - Expanded softmax benchmarks to compare standard, upstream accelerated, and VibeCUDA implementations.
  - Added scalar-temperature PDL coverage and improved performance summaries and visualizations.
- **Bug Fixes**
  - Unsupported backend selections now report a clear validation error.

### CODE REVIEW

**Comment by Edenzzzz** [comment]
@flashinfer-bot run
